Embedded/DBEngine.cpp

```
@@ -502,40 +491,56 @@ class DBEngineImpl : public DBEngine {
-      g_enable_columnar_output = std::stoi(value);
-    } else if (key == "enable_union") {
-      g_enable_union = std::stoi(value);
-    } else if (key == "enable_debug_timer") {
-      g_enable_debug_timer = std::stoi(value);
-    } else if (key == "enable_lazy_fetch") {
-      g_enable_lazy_fetch = std::stoi(value);
-    } else if (key == "udf_filename") {
-      udf_filename = value;
-    } else if (key == "null_div_by_zero") {
-      g_null_div_by_zero = std::stoi(value);
-    } else {
-      std::cerr << "WARNING: ignoring unknown DBEngine parameter '" << key << "'"
-                << std::endl;
+DBEngine* DBEngine::init(const std::string& path, int port) {
+  auto engine = DBEngineImpl::get();
+  if (!engine) {
+    g_enable_union = false;
+    g_enable_columnar_output = true;
+    engine = DBEngineImpl::create(path, port, "");
+  }
+  return engine;
+}
+
+DBEngine* DBEngine::init(const std::map<std::string, std::string>& parameters) {
+  auto engine = DBEngineImpl::get();
+  if (!engine) {
+    int port = DEFAULT_CALCITE_PORT;
+    std::string path, udf_filename;
+    g_enable_union = false;
+    g_enable_columnar_output = true;
+    for (const auto& [key, value] : parameters) {
+      if (key == "path") {
+        path = value;
+      } else if (key == "port") {
+        port = std::stoi(value);
+      } else if (key == "enable_columnar_output") {
+        g_enable_columnar_output = std::stoi(value);
+      } else if (key == "enable_union") {
+        g_enable_union = std::stoi(value);
+      } else if (key == "enable_debug_timer") {
+        g_enable_debug_timer = std::stoi(value);
+      } else if (key == "enable_lazy_fetch") {
+        g_enable_lazy_fetch = std::stoi(value);
+      } else if (key == "udf_filename") {
+        udf_filename = value;
+      } else if (key == "null_div_by_zero") {
+        g_null_div_by_zero = std::stoi(value);
+      } else {
+        std::cerr << "WARNING: ignoring unknown DBEngine parameter '" << key << "'"
```

One we have a good way to handle initialization errors, we can also return an error for unknown parameters.

---

return false added for unknown parameters